### ISSUE TITLE
tlsContext interferes with filterChainMatch and TLS inspector

### ISSUE THREAD

**Issue by vadimeisenbergibm** [envoyproxy/envoy#4076]
The scenario: an application sends TLS traffic to https://edition.cnn.com, with SNI. The traffic is intercepted by a sidecar Envoy, which sends the traffic to a gateway Envoy, which, in turn, sends the traffic to edition.cnn.com. The gateway Envoy reports the SNI.

application (SNI = edition.cnn.com) -> sidecar proxy (SNI = edition.cnn.com)-> gateway proxy  (SNI = edition.cnn.com) -> edition.cnn.com

Without mTLS between the proxies, the configuration is straightforward:  the sidecar proxy has a filter chain with a filter chain match by `server_names: edition.cnn.com`, a TCP proxy filter that forwards the traffic to the gateway. The gateway has a filter chain with a filter chain match by `server_names: edition.cnn.com`, with a TCP proxy filter that forwards the traffic to edition.cnn.com. The gateway has also a filter that calls `read_callbacks_->connection().requestedServerName()` and prints the SNI equal to `edition.cnn.com`. Everything works as expected.

Now suppose we want to apply mTLS between the sidecar proxy and the gateway proxy (so we will have an mTLS "tunnel" between the proxies, with the original TLS with the original SNI inside the tunnel). The mTLS could be required for verifying the identity of the application (by verifying the identity of the sidecar proxy).

So the picture becomes as follows:
application (SNI = edition.cnn.com) -> sidecar proxy (SNI = gateway.cluster.local)-> gateway proxy  (SNI = edition.cnn.com) -> edition.cnn.com

Note that now mTLS interferes with the filterChainMatch and SNI reporting of the gateway proxy. The gateway proxy will have to specify `server_names: gateway.cluster.local` and it will report SNI as `gateway.cluster.local`.

As far as I understand, mTLS between the proxies is transparent with regard to the other protocols, e.g. HTTP. ***The desired operation*** would be: Envoy will perform TLS termination to terminate mTLS between the proxies, and then will perform the SNI matching by the original SNI and reporting of the original SNI (in this case `edition.cnn.com`).

**Comment by vadimeisenbergibm** [comment]
cc @mattklein123 @PiotrSikora @rshriram

**Comment by mattklein123** [comment]
@PiotrSikora will know the details, but I'm not sure this is possible without using/inventing some type of TLS extension to forward the original SNI.

**Comment by PiotrSikora** [comment]
Traffic inspection / filter chain matching works on the traffic it sees (i.e. outer layer of your double-encrypted tunnel, i.e. mTLS), but you're trying to match on the inner layer, which is encrypted and thus invisible to traffic inspector. At the stage of matching, we don't even know if the filters will want to (or even can) decrypt the traffic, so there is no access to values from inner layer.

But, is there any reason why you can't keep SNI as `edition.cnn.com` for the outer layer?

**Comment by vadimeisenbergibm** [comment]
@PiotrSikora I am confused by the fact that HTTP routing and reporting works successfully on the inner layer (HTTP inside the mTLS tunnel). Envoy is able to perform TLS termination, and after that to perform routing by the decrypted Host header and to report the decrypted headers. Why the same cannot be performed for TLS inside mTLS tunnel? How TLS inside mTLS tunnel is different from HTTP inside mTLS tunnel, from Envoy's point of view?

> But, is there any reason why you can't keep SNI as edition.cnn.com for the outer layer?

Because the sidecar proxy may not know the SNI. Consider a case it is configured to forward the traffic of `*.cnn.com`, or `*.com` or `*` to the gateway. The sidecar proxy is configured with the corresponding `filterChainMatch` and it manages to forward the traffic to the gateway proxy. However, the sidecar proxy does not know the original SNI, so it cannot send the original SNI to the gateway proxy as part of the mTLS tunnel.

**Comment by elevran** [comment]
@PiotrSikora it is true that the payload is encrypted and thus invisible to the traffic inspector. However, the SNI extension part of the client handshake is sent in clear text and is therefore available. 
The external (mTLS) tunnel carries the destination service name in the SNI (e.g., `egress-gw.cluster.local`). I don't know if we can manipulate SNI on the sender side to have other values. The `edition.cnn.com` is part of the payload.

It seems that the goal is to extract it from the first SSL record in the mTLS connection payload (i.e., a traffic inspection filter) and set it as a request or connection attribute. Once that value is extracted the filter's work is done. The attribute ("inner_server_name") can be sent to the mixer for checks and possibly used to determine the upstream cluster for connection routing.

**Comment by PiotrSikora** [comment]
> @PiotrSikora I am confused by the fact that HTTP routing and reporting works successfully on the inner layer (HTTP inside the mTLS tunnel). Envoy is able to perform TLS termination, and after that to perform routing by the decrypted Host header and to report the decrypted headers. Why the same cannot be performed for TLS inside mTLS tunnel? How TLS inside mTLS tunnel is different from HTTP inside mTLS tunnel, from Envoy's point of view?

Filter chain selection happens before TLS termination, without access to the data inside the mTLS tunnel.

HTTP routing happens inside HTTP connection manager, after the traffic is already decrypted.

You could create TLS filter that does routing / cluster selection based on the TLS ClientHello message from the inner layer, after the traffic is already decrypted, but that would need to happen after filter chain is selected, because you need to know which certificate to use to complete the TLS handshake and decrypt the traffic. This is definitely doable, but requires some work.

> @PiotrSikora it is true that the payload is encrypted and thus invisible to the traffic inspector. However, the SNI extension part of the client handshake is sent in clear text and is therefore available.

SNI in the outer layer (i.e. mTLS tunnel), yes, but not the SNI in the inner layer (i.e. workload-to-workload TLS inside mTLS tunnel), which this issue is about.

> It seems that the goal is to extract it from the first SSL record in the mTLS connection payload (i.e., a traffic inspection filter) and set it as a request or connection attribute. Once that value is extracted the filter's work is done. The attribute ("inner_server_name") can be sent to the mixer for checks and possibly used to determine the upstream cluster for connection routing.

Note that filter chain selection, cluster routing and information extraction for the logging purposes are all very different things, and what you're talking about is unrelated to filter chain selection.

**Comment by vadimeisenbergibm** [comment]
> You could create TLS filter that does routing / cluster selection based on the TLS ClientHello message from the inner layer, after the traffic is already decrypted, but that would need to happen after filter chain is selected, because you need to know which certificate to use to complete the TLS handshake and decrypt the traffic. This is definitely doable, but requires some work.

OK, let's focus on this. I think what happened here is that filter chain selection for the mTLS tunnel is mixed with TLS routing of the original traffic. I think we have two options to solve it:

Option 1: Is it possible to perform filter chain selection for mTLS tunnel, then mTLS termination, and then to perform the second round of filter chain selection based on the original traffic? This will work the same with the filter chain matches with or without mTLS tunnel. You create filter chain matches for both `server_names: gateway.cluster.local` and `server_names: edition.cnn.com`. The first one performs TLS termination, the second one does not perform it. If the mTLS is used with SNI `gateway.cluster.local`, the first filter chain is selected based on the mTLS SNI, TLS termination is performed, then in the second round of filter chain selection, the second filter chain is selected based on the inner SNI,  so the routing based on the original SNI is performed and the original SNI is reported.

Option 2: to create `envoy.tls_proxy` filter based on the TLS ClientHello message from the inner layer. It will perform routing based on the inner SNI and will report the inner SNI.

Which option would be more clear for the users and more consistent? Which option would be easier to implement?

**Comment by rshriram** [comment]
I think you should create a new filter in both http and tcp. The filter needs to read the incoming server_name field and stick this into the RequestInfo object. Then in the cluster, you need to check if the user has explicitly set the SNI field in the cluster. if yes, you use the user provided value. If no, you look at the RequestInfo object to see if any filter has set the SNI value, and use it. 

This will allow you to propagate the incoming SNI value from app as outbound sni value in the sidecar to gateway..

**Comment by vadimeisenbergibm** [comment]
@rshriram So you propose option 3: the sidecar proxy will extract the SNI and forward it as the SNI of the mTLS tunnel.

@PiotrSikora @mattklein123 So which option among the options in https://github.com/envoyproxy/envoy/issues/4076#issuecomment-411616822, https://github.com/envoyproxy/envoy/issues/4076#issuecomment-411620562 is better? Which one is easier to implement?

**Comment by elevran** [comment]
@vadimeisenbergibm based on my understanding the purpose of egress service is to protect against cases where the sending proxy has been compromised, meaning we don't want to trust the SNI it sent but rather extract the client SNI from the inner TLS (sine that's what we'll use as the target).

@PiotrSikora:
> SNI in the outer layer (i.e. mTLS tunnel), yes, but not the SNI in the inner layer (i.e. workload-to-workload TLS inside mTLS tunnel), which this issue is about.

I may not be correctly understanding your comment. The inner SNI is encrypted using the mTLS. I was suggesting to extract the inner SNI in the receiving proxy, after it decrypts the mTLS connection. At that point the inner SNI is clear text again and can be inspected.

**Comment by vadimeisenbergibm** [comment]
@elevran Right, the option 3 would not help us to prevent cheating (sending a fake SNI) by a compromised sidecar proxy.

**Comment by rshriram** [comment]
How is this improving security? is someone MITMing and changing the SNI alone? that can't happen in a k8s network as there is no L2. Is someone masquerading as the source pod? That can't happen if you do RBAC at the egress gateway that validates against the source pod's SAN. 

Doing the double/triple mtls inspection stuff is something that cannot be easily programmed through Pilot as it treats ingress and egress gateways in the same way.  This will then end up manifesting as a multitude of user configurations (classify gateway type or define a special TLS mode, etc. etc.). The gateway has no idea whether the payload inside a mTLS connection is already encrypted or not or its protocol etc. So, simply reading the SNI string verbatim is a security hole in itself as the protocol is not typed [we don't know what it is in advance].

**Comment by vadimeisenbergibm** [comment]
The scenario is as follows. Suppose you have the following policy in Istio:

1. PodA may access *.foxnews.com
2. PodB may access *.cnn.com

An attacker is able to break into the PodA and wants to access edition.cnn.com, which is forbidden by the example policy.

Let's assume that the attacker is able to kill/bypass the sidecar proxy, steal its certificates (stealing certificates is mentioned here - https://istio.io/docs/concepts/security/#secure-naming) and to run its own, malicious version of the sidecar proxy with the certificates of the PodA. Now the attacker will issue a request to edition.cnn.com (with the original SNI equal `edition.cnn.com`). The malicious sidecar proxy will open mTLS connection to the gateway, with the outer SNI equal `www.foxnews.com`. The gateway's proxy will send to the Istio Mixer the identity of the pod `PodA` and the outer SNI `www.foxnews.com`. The Mixer check will succeed. Then the gateway will perform mTLS termination, and will forward the request by the original SNI to `edition.cnn.com` (see https://github.com/istio/istio.github.io/pull/1984).

To prevent such attack we need to ensure that the gateway sends the original, inner SNI to the mixer, not the outer one. The security is enforced in the following way - you forward the traffic to the same server that you report to the mixer. This way the traffic will indeed be forwarded to the server that was reported to and allowed by the Mixer.

**Comment by elevran** [comment]
@rshriram do you think it could, perhaps, be cleaner if we had a clear separation between tunnel setup phase (where envoys can exchange any data needed - e.g., to support mTLS, forward original client a server L3 information, etc.) and payload exchange after the tunnel is fully set up?

This would work well for TCP proxying where the envoy-envoy connection is "per session", and may be skipped for when multiplexing multiple connections unto the same envoy connection. SNI routing is more like TCP proxying than HTTP/H2.

**Comment by rshriram** [comment]
I think tunelling is the way to go. But rather than doing a TCP tunnel, we should tunnel  over h2. The facility exists (almost) in the HTTP connection manager - being used currently for Websockets. Having H2 tunnels would allow sending more metadata.

**Comment by ggreenway** [comment]
Sorry I'm chiming in late (was on vacation). 

Could you solve this by, in the handling of the outer (tunnel) TLS connection, route to a cluster that connects to another envoy listener? This 2nd listener could be on localhost, or a unix-domain socket. Then you can use the existing tls_inspector to do your routing on the inner TLS connection.

**Comment by vadimeisenbergibm** [comment]
@ggreenway The question is the performance impact of this solution. Is it like an additional hop on the localhost? Or will the additional "hop" happen in the memory of the Envoy process?

**Comment by ggreenway** [comment]
It would be an extra hop on localhost. If it is through a UDS, that will lessen the cost.

In theory, Envoy could implement some sort of VIP redirection mode that does this in memory, but that would require some careful thought on how it is designed.

**Comment by vadimeisenbergibm** [comment]
@ggreenway Thank you for the idea, it is good that we have this option as a fallback, however we want to save the extra hop on localhost.

**Comment by lizan** [comment]
> option 3: the sidecar proxy will extract the SNI and forward it as the SNI of the mTLS tunnel.

+1 to @rshriram's option 3, and avoid #4236, loopback cluster to another listener is slightly better than that.

> @vadimeisenbergibm based on my understanding the purpose of egress service is to protect against cases where the sending proxy has been compromised, meaning we don't want to trust the SNI it sent but rather extract the client SNI from the inner TLS (since that's what we'll use as the target).

@elevran The sidecar (sending proxy) has same security boundary within the client, so if it is compromised, egress should treat both client and sending proxy is compromised. Sending proxy could craft a TLS traffic within a mTLS tunnel and send to egress too.

**Comment by vadimeisenbergibm** [comment]
@lizan I have two problems with the loopback cluster:

1. Wasting performance
2. The identity of the source will be lost, the second listener will report the inner SNI value correctly, but it will not have the source identity to report. I need a mixer check that will report both SNI value and the identity of the source, to apply a mixer policy based on both of them.

**Comment by htuch** [comment]
I wonder if we can loopback and include the outer SNI in a header for filters to work on for (2). I agree that there is some performance hit here for (1); is this a specific concern for your application?

If we do want to go the route in this PR, I think trying to factor out and share as much as possible with the transport socket SNI handling makes sense.

**Comment by vadimeisenbergibm** [comment]
@htuch  
>  if we can loopback and include the outer SNI in a header for filters to work on for

Isn't it a more complex solution comparing to the simple filter we propose in #4236 ? The requirement is very simple - we need the value of the inner SNI (the SNI that was used by the application) to be accessible to other filters in a simple way. Here we started a discussion about the scenarios this inner SNI is needed in, which I am not sure are really relevant here. My guess is that TLS is one of the most important protocols and its monitoring is also one of the important features for Envoy, in my opinion. Currently, with mTLS between two Envoys, the value of the inner SNI cannot be monitored.

The solution is pretty simple as well - adding a simple filter (currently its .cc file is less than 90 lines), let it extract the inner SNI and stick it somewhere, for example in the ReaderFilterCallbacks, so other filters could read this value.

**Comment by rshriram** [comment]
Also keep in mind that in your scenario ()https://github.com/envoyproxy/envoy/issues/4076#issuecomment-412322855), if you assume the sidecar is compromized, it could send any arbitrary junk TLS traffic in the inner TLS, and stick any SNI it wants. There is also nothing preventing the compromised sidecar from modifying the SNI set by the client before forwarding it inside the mTLS wrapper. SNI is in clear text. Its not signed/protected.

**Comment by vadimeisenbergibm** [comment]
@lizan

> No the gateway will forward inner SNI edition.cnn.com to the cluster matches www.foxnews.com

I need to present the full solution here. Please see this PR https://github.com/istio/istio.github.io/pull/1984, in particular this explanation:

> Configuring traffic to arbitrary wildcarded domains introduces a challenge for Istio gateways. In the previous section
you directed the traffic to _www.wikipedia.org_, and this host was known to your gateway during the configuration.
The gateway, however, cannot know an IP address of an arbitrary host it receives a request for. Would you want to
control access to `*.com`, and send requests to _www.cnn.com_ and _www.abc.com_, the Istio gateway would not know which
IP address to forward the requests.
This limitation is due to the limitation of [Envoy](https://www.envoyproxy.io), the proxy Istio is based on. Envoy
routes traffic either to predefined hosts, or to predefined IP addresses, or to the original destination IP address of
the request. In the case of the gateway the original destination IP of the request is lost (since the request was routed
to the egress gateway and its destination IP address is the IP address of the gateway).

> In short, the Istio gateway based on Envoy cannot route traffic to an arbitrary, not preconfigured host, and AS-IS is
unable to perform traffic control to arbitrary wildcarded domains. To enable such traffic control for HTTPS (and for any
TLS), you need to deploy an SNI forward proxy in addition to Envoy. Envoy will route the requests destined for a
wildcarded domain to the SNI forward proxy, which, in turn, will forward the requests to the destination by the value of
SNI.

In our solution, Envoy on the gateway forwards the traffic to an additional, SNI proxy on the local host. The SNI proxy then forwards the traffic by the inner SNI. So in the case of the attack described in https://github.com/envoyproxy/envoy/issues/4076#issuecomment-412322855, the attack will succeed since the traffic will go by the inner SNI, which is `edition.cnn.com`.

I propose the following principle: the routing must be performed by the same SNI, which is reported to the Mixer. Currently, Mixer is reported the outer SNI while the traffic is routed (using the SNI proxy) by the inner SNI.

**Comment by vadimeisenbergibm** [comment]
@rshriram 
>  if you assume the sidecar is compromized, it could send any arbitrary junk TLS traffic in the inner TLS, and stick any SNI it wants. 

So I need istio-proxy to report that inner SNI with junk SNI to the Mixer, so the connection will be blocked by a Mixer policy, and the junk SNI will appear in the Mixer logs.

> There is also nothing preventing the compromised sidecar from modifying the SNI set by the client before forwarding it inside the mTLS wrapper

Right, this is why I need that inner SNI to be reported to the Mixer, for the policy checking and for monitoring. With the current Envoy, I cannot know the value of the inner SNI.

See also https://github.com/envoyproxy/envoy/issues/4076#issuecomment-416097927

**Comment by lizan** [comment]
> In our solution, Envoy on the gateway forwards the traffic to an additional, SNI proxy on the local host. The SNI proxy then forwards the traffic by the inner SNI. So in the case of the attack described in https://github.com/envoyproxy/envoy/issues/4076#issuecomment-412322855, the attack will succeed since the traffic will go by the inner SNI, which is `edition.cnn.com`.

Why do you need another SNI proxy on localhost to forward the traffic with inner SNI, instead of using outer SNI directly? Note not all inner TLS have SNI if you assume it could be some legacy clients.

**Comment by vadimeisenbergibm** [comment]
@lizan 

The use case: enable HTTPS egress traffic to `*.ibm.com` and `*.amazon.com` through an egress gateway. Requests to `*.ibm.com` and to `*.amazon.com` must be allowed and monitored, and policies must be enforced on them, for example, `podA` may access `service1.ibm.com`, while `podB` may access `service2.amazon.com`, podC may access `*.ibm.com`, all other access is forbidden.

How can you configure Envoy to perform the routing according to the outer SNI?

**Comment by lizan** [comment]
@vadimeisenbergibm If your sidecar extracts inner SNI and forward it as outer SNI (option 3), mixer can enforce them. The routing is what you do with the another SNI proxy on localhost described above (i.e. by filter chain in the listener receiving outer SNI).

**Comment by vadimeisenbergibm** [comment]
Discussing it with @rshriram on slack. How can Envoy in the gateway set the outer SNI to be the SNI of the cluster that forwards the traffic to the SNI proxy, so the SNI proxy will receive the outer SNI? You can have a filter that reads the value of the outer SNI, but can that filter modify the cluster and set the outer SNI to be the SNI of the cluster? Can filters modify clusters?

**Comment by vadimeisenbergibm** [comment]
@lizan 
> If your sidecar extracts inner SNI and forward it as outer SNI

How can Envoy set the SNI of the cluster? How can it forward the extracted SNI (unknown beforehand) to the gateway?

**Comment by lizan** [comment]
Right, that part doesn’t exist today so you need to implement it in tcp_proxy to forward same SNI inspected by the inspector.

That seems a more generic feature than the approach in #4236. It will allow tcp_proxy to forward original SNI if TLS termination is performed, and also support the tunneling case described here if you only inspection. @PiotrSikora @ggreenway?

**Comment by lizan** [comment]
Sorry just accidentally hit wrong button :(

**Comment by vadimeisenbergibm** [comment]
@lizan 
> Right, that part doesn’t exist today so you need to implement it in tcp_proxy to forward same SNI inspected by the inspector.

cc @rshriram 

OK, great, so we agree that our requirement cannot be implemented today with Envoy AS IS.
The requirement is pretty straightforward - to extract the inner SNI by a read filter and stick it somewhere in the Envoy's data structures so other filters will be able to read it. The implementation is also very simple.

You propose a more complex change - `tcp_proxy` will get the outer SNI and it will interfere with the original TLS and replace the inner SNI with the outer SNI.

So why don't we just proceed with a simple filter that implements a straightforward requirement?

**Comment by vadimeisenbergibm** [comment]
@lizan @rshriram 

> If your sidecar extracts inner SNI and forward it as outer SNI

This part I am still missing. Can a filter in Envoy set the SNI of a cluster? In the case of the Envoy in the sidecar, we have a cluster that opens mTLS to the gateway. So can a filter change mTLS settings of a cluster?

**Comment by lizan** [comment]
> OK, great, so we agree that our requirement cannot be implemented today with Envoy AS IS.
The requirement is pretty straightforward - to extract the inner SNI by a read filter and stick it somewhere in the Envoy's data structures so other filters will be able to read it. The implementation is also very simple.

Yes, I think all options listed above need some code change in Envoy, except tunneling to another listener which sacrifices performance.

I would echo @rshriram's comment that the approach of #4236 takes is not a useful long term solution for the problem it is tackling. 

> You propose a more complex change - `tcp_proxy` will get the outer SNI and it will interfere with the original TLS and replace the inner SNI with the outer SNI.

No, that's not I'm suggesting, I'm suggesting in sidecar Envoy, extract the SNI from client side (inner SNI) and forward it as outer SNI for mTLS.

**Comment by vadimeisenbergibm** [comment]
@lizan 
> I'm suggesting in sidecar Envoy, extract the SNI from client side (inner SNI) and forward it as outer SNI for mTLS.

How can a filter set SNI for Envoy's mTLS?

**Comment by vadimeisenbergibm** [comment]
@rshriram @lizan @PiotrSikora I decided to document the requirements, the design and the implementation issues for the whole use case in a design proposal In Istio https://goo.gl/cQ29rX.

Let's discuss the whole use case first. While we collected this use case from several customers, and one customer needs this implementation urgently, I think the use case is very straightforward and totally generic. Please see the Requirements section.

I would like to repeat that #4236 is just a small simple filter that reads the SNI and stores it in `ReadFilterCallbacks`. It fills a piece of missing functionality in Envoy, described in the initial comment and can be understood out of the broader context (the use case described in https://goo.gl/cQ29rX).

While one of our customers needs this functionality urgently, they are not ready to adopt a custom version of Envoy or of Istio-proxy.

**Comment by cmluciano** [comment]
FYI I added REQUESTED_SERVER_NAME within #4144 to the request_info. I'm not sure if that helps as far as populating a variable to pass on, but we at least are able to record the values in the access_log now that this is merged.

**Comment by vadimeisenbergibm** [comment]
@cmluciano Thanks, Chris! It is definitely very useful.

**Comment by lizan** [comment]
> How can a filter set SNI for Envoy's mTLS?

Need to change `tcp_proxy` when it creates connection to upstream, pass the SNI, and plumbing down to here:
https://github.com/envoyproxy/envoy/blob/master/source/common/ssl/context_impl.cc#L493

> I would like to repeat that #4236 is just a small simple filter that reads the SNI and stores it in `ReadFilterCallbacks`.

The SNI is already available in `ReadFilterCallbacks` without the filter proposed in #4236, i.e. read_callbacks_->connection().requestedServerName().

**Comment by rshriram** [comment]
I second @lizan suggestion. Making the tcp proxy forward the SNI from inbound connection to outbound cluster is a far more generic solution than the PR you have. If another customer comes with the requirement that inner tls SNI has to be stripped, and decides to add a filter to Envoy open source, what would you do ? 

I also don’t see a clean way to surface this in istio without adding a myriad of api options for this very specific use case.

**Comment by cmluciano** [comment]
> Need to change tcp_proxy when it creates connection to upstream, pass the SNI, and plumbing down to here:
https://github.com/envoyproxy/envoy/blob/master/source/common/ssl/context_impl.cc#L493

@vadimeisenbergibm do you think ^ will not be sufficient?

**Comment by lizan** [comment]
Solution B sounds better, though I don't think you really want to do 2, at egress you can just use outer SNI for both mixer check and routing, then you don't have to worry about inner SNI. It doesn't add much value as long as you don't perform routing based on inner SNI: https://github.com/envoyproxy/envoy/issues/4076#issuecomment-416342017 

> 1. Change `tcp_proxy` to set the SNI of the upstream connection in case the cluster's tls context does not have `sni` field.

Add explicit field `forward_original_sni` would be better than defaulting to forward.

**Comment by vadimeisenbergibm** [comment]
@cmluciano @lizan Note that the case of Envoy in the sidecar proxy is different from the case of Envoy in the gateway, with regard to the Solution B.

The sidecar's `tcp_proxy` forwards the traffic to a cluster with `tls_context` (the mTLS). `tcp_proxy` has to fetch the `tls_context` of the upstream connection and set the SNI in it.

The gateway's `tcp_proxy` forwards the traffic to a cluster without `tls_context`. It just forwards the original TLS as opaque TCP. It is not so simple to set the SNI in this case. Here `tcp_proxy` has to interfere with the original traffic that contains the original SNI - it has to buffer the client hello message, replace the original SNI with the outer SNI, and forward the traffic. It seems it is simpler to perform Solution B, step 2.

**Comment by lizan** [comment]
> Here `tcp_proxy` has to interfere with the original traffic that contains the original SNI - it has to buffer the client hello message, replace the original SNI with the outer SNI, and forward the traffic. It seems it is simpler to perform Solution B, step 2.

Can you clarify why it has to do them? I think I made it clear you need neither. Just forward as opaque, routing based on outer SNI should suffice.

**Comment by vadimeisenbergibm** [comment]
@lizan 
> routing based on outer SNI should suffice

Two points:
1. Mixer must be reported the host the request will be routed to. Mixer also must be reported the host that will be sent to the destination as the value of the SNI (both values should be the same). Even if Envoy could perform the routing according to the outer SNI, the inner SNI could be different, another discrepancy between what Mixer is reported and what happens in reality.
2. Envoy cannot route to an arbitrary host. Suppose we allow the traffic to `*.foo.com` and a request arrives to `mongodb1.foo.com`. Envoy cannot have a cluster with every host of `*.foo.com` configured. Envoy does not know how to route to `mongodb1.foo.com`, since it has no cluster for it. To overcome this limitation of Envoy, we direct the request to the SNI proxy that forwards the request by **the inner SNI**.

The following attack is possible if we just direct the traffic to the SNI proxy by the outer SNI:
 
The policies: All the microservices are allowed to access `*.foo.com`, except the pod _foo_ that is allowed to access `mongodb2.foo.com` only. We configure a filter chain for `*.foo.com`, and add a Mixer policy that checks the source principal and the SNI.

The attack:
The request will be sent by a fake proxy with inner SNI `mongodb1.foo.com` and the outer SNI `mongodb2.foo.com`. The request will pass `*.foo.com` filter chain, the mixer will get the outer SNI `mongodb2.foo.com`, return `allow`, the `tcp_proxy` will forward the request to the SNI proxy that will perform the routing by the inner SNI to the forbidden for foo host `mongodb1.foo.com`.

**Comment by lizan** [comment]
> 2. Envoy cannot route to an arbitrary host. Suppose we allow the traffic to `*.foo.com` and a request arrives to `mongodb1.foo.com`. Envoy cannot have a cluster with every host of `*.foo.com` configured. Envoy does not know how to route to `mongodb1.foo.com`, since it has no cluster for it. To overcome this limitation of Envoy, we direct the request to the SNI proxy that forwards the request by **the inner SNI**.

This is a workaround, the longer term solution is add that ability to Envoy and use outer SNI to route, not sure if the architecture of adding another SNI proxy is the right way. I'll comment more in your design doc for istio.

> The attack:
The request will be sent by a fake proxy with inner SNI `mongodb1.foo.com` and the outer SNI `mongodb2.foo.com`. The request will pass `*.foo.com` filter chain, the mixer will get the outer SNI `mongodb2.foo.com`, return `allow`, the `tcp_proxy` will forward the request to the SNI proxy that will perform the routing by the inner SNI to the forbidden for foo host `mongodb1.foo.com`.

Fake (compromised) proxy is the problem that mTLS identity layer should address (i.e. revoke the sidecar certificate / re-issue), and you can only do your best here. It is already pointed out in https://github.com/envoyproxy/envoy/issues/4076#issuecomment-416023542 and https://github.com/envoyproxy/envoy/issues/4076#issuecomment-416081280 that fake proxy can send arbitrary inner SNI that intended to make egress trust it, so even you do the verify, as long as the route is happening based on outer SNI (opposed to above) and it is used for mixer, then the problem is mitigated.

More over, you don't really have the control of where the request go, you cannot inspect the `:authority` or `Host` header inside the HTTPS traffic, that will still allow fake proxy to generate traffic to unintended host. For some real example, just try run `curl -vvvv https://ga-dev-tools.appspot.com/ -X HEAD -H "Host: k8s-gubernator.appspot.com"` and you will see SNI is former but you get contents from latter. In most of the cases you want do wildcard SNI this is true.

**Comment by rshriram** [comment]
How will you surface this as configuration in Istio? How will Pilot know whether to install the SNI reader filter or not for a given port on a proxy? We do not (and should not) be differentiating ingress/egress gateways. A gateway is simply a gateway. 

This whole workaround is being done for the missing capability where envoy cannot send to arbitrary DNS destinations. Once that feature lands, this whole thing will become an ugly workaround and the code will rot, just like the validate clusters option in route (which we asked for and we don't use). Not to mention, any code added to Istio will also rot and we will end up having to put up with backward compatibility issues. 

Which is why I suggested that you maintain this filter out of tree and deploy a custom istio proxy for the customer so that when the DNS feature lands, they can switch to the upstream istio-proxy instead of imposing this burden on everyone else in the community.

**Comment by vadimeisenbergibm** [comment]
@rshriram 

> How will you surface this as configuration in Istio? How will Pilot know whether to install the SNI reader filter or not for a given port on a proxy? 

By specifying EnvoyFilter https://istio.io/docs/reference/config/istio.networking.v1alpha3/#EnvoyFilter-Filter

The customer will specify SNIReader's location by specifying listener's port, port name and the type `GATEWAY`. The customer can choose some convention for the port names, e.g. tis-egress-443 and apply the filter on these ports.

> This whole workaround is being done for the missing capability where envoy cannot send to arbitrary DNS destinations. Once that feature lands, this whole thing will become an ugly workaround and the code will rot

Once this feature lands, I will be the first to remove the SNI proxy and simplify the code. Everybody will be happy to have a single proxy instead of two, complexity- and performance-wise.

>  you maintain this filter out of tree and deploy a custom istio proxy for the customer

The customers may decide that they do not want a custom Istio proxy since they do not want to provide support for their custom Istio proxy. This will lessen the attractivity of Istio as an open-source solution for the customer, once the customer has to support parts of it.

Again: we propose to add to Envoy a small filter that will extract the inner SNI and stick it to ReadFilterCallbacks. This filter implements a small piece of missing functionality described in the first comment of this issue. We also propose to add to Istio-proxy `mixer` filter sending this one additional attribute to Mixer.

**Comment by vadimeisenbergibm** [comment]
@lizan 

> This is a workaround, the longer term solution is add that ability to Envoy and use outer SNI to route, not sure if the architecture of adding another SNI proxy is the right way

Once the long term solution is available in Envoy, I will be the first to remove the SNI proxy and simplify the code.

> Fake (compromised) proxy is the problem that mTLS identity layer should address (i.e. revoke the sidecar certificate / re-issue),

Citadel cannot know if there is some fake proxy running in a pod. The fake proxy will steal the certificates used by the legitimate sidecar proxy in the pod.

> More over, you don't really have the control of where the request go

You have the control in the sense that the encrypted request will arrive to the hostname by the SNI. This is the requirement. The external sever specified in the SNI will decrypt the request and then decide to forward the request to any other server - this internal traffic at the external site is out of scope of the egress control.

Let's see what happens in the case you described:

`curl -vvvv https://ga-dev-tools.appspot.com/ -X HEAD -H "Host: k8s-gubernator.appspot.com"`
You will see the following line:
`Connected to ga-dev-tools.appspot.com (216.58.210.52) port 443 (#0)`. The request indeed arrived to `ga-dev-tools.appspot.com` as required. Then `ga-dev-tools.appspot.com` forwarded the request to `k8s-gubernator.appspot.com` by the Host header. That's fine, it is internal traffic at the external service site, out of scope of Istio egress traffic control.

Take for example a rule that allows `*.amazon.com` and try the following: `curl -vvvv https://www.amazon.com/ -H "Host: aws.amazon.com"`

You will get 403.

> as long as the route is happening based on outer SNI 

So we are still with the problem - how `tcp_proxy` can perform routing based on the outer SNI when its cluster does not have TLS context.

**Comment by lizan** [comment]
> You have the control in the sense that the encrypted request will arrive to the hostname by the SNI. This is the requirement. The external sever specified in the SNI will decrypt the request and then decide to forward the request to any other server - this internal traffic at the external site is out of scope of the egress control.

From that point, I'd argue inner SNI is also out of scope of egress control, what about if the application doesn't use SNI? It is receiving server's responsibility to verify, as long as you don't route based on it.

>  Then ga-dev-tools.appspot.com forwarded the request to k8s-gubernator.appspot.com by the Host header.

This is definitely forwarding, but a common behavior for multitenancy or dynamic hosting has, (which is where you really want to do wildcard without knowing what the host is).
```
curl -v https://istio.github.io/ -H "Host: envoyproxy.io"
curl -v https://istio.github.io/ -H "Host: envoyproxy.github.io.io"
```
will give you redirect and envoy content.

**Comment by lizan** [comment]
I think we all agreed in longer term:
- a generic outbound proxy https://github.com/envoyproxy/envoy/issues/1606
- tunneling over HTTP https://github.com/envoyproxy/envoy/issues/1451
along with the metadata of SNI.

In short term, I think having mixer filter to inspect inner SNI is acceptable, though I don't think the data itself is as generic as to be be stored in `ReadFilterCallbacks`.

**Comment by vadimeisenbergibm** [comment]
@lizan 
> what about if the application doesn't use SNI

The requirements we received from our customers is to support TLS with SNI. We were not required to support TLS without SNI.

**Comment by vadimeisenbergibm** [comment]
@lizan 

> In short term, I think having mixer filter to inspect inner SNI is acceptable, though I don't think the data itself is as generic as to be be stored in ReadFilterCallbacks.

OK, agreed for the short term. So where should we store the inner SNI?

**Comment by lizan** [comment]
You can do that in mixer filter alone (extracting and verifying) once we agreed on the proposal in istio community to do the short term instead of directly pursuing long term.

**Comment by vadimeisenbergibm** [comment]
@lizan OK, we will perform only the refactoring of tls inspector in Envoy, and add the extracting and verifying functionality to mixer.

**Comment by elevran** [comment]
regarding suggestion by @rshriram 
> This will allow you to propagate the incoming SNI value from app as outbound sni value in the sidecar to gateway..

and @lizan: 
> No, that's not I'm suggesting, I'm suggesting in sidecar Envoy, extract the SNI from client side (inner SNI) and forward it as outer SNI for mTLS.

Could you please clarify the proposed solution?

My understanding of the current flow is as follows:
1. client uses (for example) `target-server.example.org` as the (inner) SNI value;
1. traffic is intercepted by and then encapsulated by client envoy proxy and sent to egress service with (outer) SNI value of (e.g.) `istio-egress`; 
1. egress proxy needs to extract the inner SNI value (`target-server.example.org`) and route the request to it

Is the suggestion to sidestep the last step by having the client proxy use an (outer) SNI value copied from the client (inner) SNI? That is, both inner and outer would be set to `target-server.example.org` in step 2, between the client envoy proxy and the egress service envoy proxy, so there is no need to extract the inner SNI value for routing in the last step.
